Add hooks to override protocol version numbers

rnicoll opened the pull request:
Enables altcoin networks which may have out of sync protocol numbers, and also moves version numbers into a single enum rather than having them scattered through the code.

schildbach commented on core/src/main/java/org/bitcoinj/core/NetworkParameters.java:
```diff
@@ -505,4 +500,23 @@ public int getMajorityWindow() {
+
+        private int bitcoinProtocol;
+
+        ProtocolVersion(final int setNum) {
```
"setNum" feels like a weird name for a method parameter. Maybe "bitcoinProtocol"?

schildbach commented:
Code looks ok.

I realize one of the PeerAddress constructors changed in a backwards incompatible way. I wonder if we should wait with this refactoring until the Context refactoring is complete (contexts contain network params). Though I'm not sure if the Context refactoring is planned to touch PeerAddress, maybe @mikehearn knows?

rnicoll commented:
Changed the constructor variable name, and fixed the backwards incompatible change in PeerAddress. May want to deprecate some of the PeerAddress constructors later, though.

rnicoll commented:
Unit tests run fine locally, and there's no obvious errors in the logs. Can you take a look and unless I've missed something, restart the build, see if that fixes it?

schildbach commented:
Merged!
